031-custom_authentication.md

```diff
@@ -0,0 +1,96 @@
+
+```
+listener.name.<listener-name>.oauthbearer.sasl.client.callback.handler.class=
+listener.name.<listener-name>.oauthbearer.sasl.server.callback.handler.class=
+listener.name.<listener-name>.oauthbearer.sasl.login.callback.handler.class=
+listener.name.<listener-name>.oauthbearer.connections.max.reauth.ms=
+listener.name.<listener-name>.sasl.enabled.mechanisms=
+listener.name.<listener-name>.oauthbearer.sasl.jaas.config=
+listener.name.<listener-name>.AWS_MSK_IAM.sasl.jaas.config=
+listener.name.<listener-name>.AWS_MSK_IAM.sasl.client.callback.handler.class=
+security.protocol=
+principal.builder.class=
+```
+
+Really, this can be boiled down to: 
+
+```
+listener.name.<lister-name>.<auth-type>.*
+listener.name.<listener-name>.sasl.enabled.mechanisms
+security.protocol
+principal.builder.class
+```
+
+Taking this further, the desired yaml should look like: 
+
+```
+    listeners:
+      - name: tls
+        port: 9093
+        type: internal
+        tls: true
+        authentication:
+          type: custom
+          principal.builder.class: SimplePrincipal.class
+          security.protocol: SASL_SSL
+          sasl.enabled.mechanisms:
+            - OAUTHBEARER
+          listener-config:
+            oauthbearer.sasl.client.callback.handler.class: client.class
+            oauthbearer.sasl.server.callback.handler.class: server.class
+            oauthbearer.sasl.login.callback.handler.class: login.class
+            oauthbearer.connections.max.reauth.ms: 999999999
+```
+
+Then, when constructing the broker config:
+
+* `principal.builder.class` is set cluster-wide for all authentication methods (this is the status quo for OAuth authentication). This can only be set once, and we will fail hard (i.e. reject the config) if there are multiple principals defined. 
+    
+    It is expected / will be documented that the provided principal builder *must* support Strimzi authentication (i.e. ssl-based auth). 
+* We parse out `security.protocol` and pass that value into the `listener.security.protocol.map` ** as we currently do for OAuth authentication types. 
+* `name` determines the prefix for all entries in `listener-config`
+    * I.e., `listener.name.<listener-name>.oauthbearer.sasl.client.callback.handler.class=`
+
+The render config, given the above example, then should look like:
+
+```
+listener.name.tls-9093.oauthbearer.sasl.client.callback.handler.class=client.class
+listener.name.tls-9093.oauthbearer.sasl.server.callback.handler.class=server.class
+listener.name.tls-9093.oauthbearer.sasl.login.callback.handler.class=login.class
+listener.name.tls-9093.oauthbearer.connections.max.reauth.ms=999999999
+principal.builder.class=SimplePrincipal.class
+listener.security.protocol.map=CONTROLPLANE-9090:SSL,REPLICATION-9091:SSL,TLS-9093:SASL_SSL
+```
+
+## Affected/not affected projects
+
+[strimzi-kafka-operator](https://github.com/strimzi/strimzi-kafka-operator)
```

Comment:
I assume that this is affected. But you should make it more clear since the title is `Affected/not affected`